model/proxy/alphav1/config/cfg.proto

```diff
@@ -21,230 +21,162 @@ package istio.proxy.v1alpha.config;
-  // Faults include aborting the Http request from downstream service,
-  // delaying the proxying of request to the upstream service, or both.
-  // TODO: This should move into UpstreamCluster
-  HttpFaultInjection fault = 3;
-
-  // Custom properties per rule (depends on proxy), such as open tracing,
-  // access log formats, etc. Some of these will be defined in future
-  // iterations.
-  google.protobuf.Any custom_impl = 4;
+  // L4 fault injection policy applies to L4 traffic
+  L4FaultInjection l4_fault = 7;
+
+  // Custom policy implementations
+  google.protobuf.Any custom = 8;
 }
 
-// Basic routing rule match criterion using Tcp attributes and downstream
-// cluster identifier
-message L4MatchCondition {
-  // Set of layer 4 match attributes such as src ip/port, dst ip/port and
-  // protocol
-  L4MatchAttributes l4attributes = 1;
+// Route rule provides a custom routing policy based on the source and
+// destination service versions and connection/request metadata.  The rule must
+// provide a set of conditions for each protocol (TCP, UDP, HTTP) that the
+// destination service exposes on its ports. The rule applies only to the ports
+// on the destination service for which it provides protocol-specific match
+// condition, e.g. if the rule does not specify TCP condition, the rule does
+// not apply to TCP traffic towards the destination service.
+message RouteRule {
+  // Uniquely identifies the destination service (not its versions) associated
+  // with this routing rule. This is applicable for hostname-based resolution for
+  // HTTP traffic as well as IP-based resolution for TCP/UDP traffic.  *required*
+  string destination = 1;
+
+  // Identifies the service initiating a connection or a request.
+  string source = 2;
+
+  // Identifies the source service version. The identifier is interpreted
+  // by the platform to match a service version for the source service.
+  //
+  // N.B. The map is used instead of pstruct due to lack of serialization support
+  // in golang protobuf library (see https://github.com/golang/protobuf/pull/208)
+  map<string, string> source_version = 3;
 
-  // Identify the downstream cluster initiating the connection.
-  ClusterIdentifier src_cluster = 2;
+  // Set of layer 4 match conditions based on the IP ranges
+  L4MatchAttributes tcp = 4;
+
+  // Set of layer 4 match conditions based on the IP ranges
+  L4MatchAttributes udp = 5;
+
+  // Set of HTTP match conditions based on the request metadata
+  HTTPMatchAttributes http = 6;
+
+  // Each routing rule is associated with one or more service versions (see
+  // glossary in beginning of document). Weights associated with the service
+  // version determine the proportion of traffic it receives.
+  repeated DestinationWeight versions = 7;
+
+  // Precedence is used to disambiguate the order of application of rules
+  // for the same destination service. A higher number takes priority.
+  int32 precedence = 8;
```

Comment:
There are no transactions in k8s, there's only optimistic concurrency. Hence, we should be dealing with revision at a higher-level than this proto. Every proto has its etcd revision number implicitly. That way we can provide optimistic apply/edit like kubctl. I don't think Manager can provide strong guarantees about rollouts though, given how k8s is structured.
The priority can be used for this too, and there's nothing wrong with that. I'll add a comment.